WaveManager/src/main/java/com/github/sef24sp4/wavemanager/WaveManager.java

```diff
@@ -0,0 +1,201 @@
+package com.github.sef24sp4.wavemanager;
+
+import com.github.sef24sp4.common.enemy.CommonEnemy;
+import com.github.sef24sp4.common.enemy.EnemyRole;
+import com.github.sef24sp4.common.enemy.EnemySPI;
+import com.github.sef24sp4.common.entities.IEntity;
+import com.github.sef24sp4.common.interfaces.IGameSettings;
+import com.github.sef24sp4.common.vector.Coordinates;
+
+import java.util.*;
+
+public class WaveManager implements IWaveManager {
+
+	private Collection<IEntity> enemyComposition = new HashSet<>();
+
+	/**
+	 * used instead of a map.
+	 */
+	private final IGameSettings gameSettings;
+
+	private int waveNumber;
+
+	private long waveDifficulty;
+
+	private final int timeUntilNextWaveInSeconds;
+
+	private long timeOfLastCheck;
+
+	private WaveStatus waveStatus = WaveStatus.ONGOING;
+
+	private Map<EnemyRole, List<EnemySPI>> enemyCatalog = new HashMap<>();
+
+	public WaveManager(int startWaveNumber, int waveCountdownInSeconds, IGameSettings gameSettings) {
+		if (startWaveNumber < 1) throw new IllegalArgumentException("startWaveNumber must be greater than 0");
+		this.waveNumber = startWaveNumber - 1; // gets +1 in the nextWave function
+		this.gameSettings = gameSettings;
+		this.timeUntilNextWaveInSeconds = waveCountdownInSeconds;
+		this.constructCatalog();
+	}
+
+	@Override
+	public WaveStatus getWaveStatus() {
+		return this.waveStatus;
+	}
+
+	@Override
+	public void setWaveStatus(WaveStatus waveStatus) {
+		this.waveStatus = waveStatus;
+	}
+
+	@Override
+	public int getWaveNumber() {
+		return this.waveNumber;
+	}
+
+	@Override
+	public long getWaveDifficulty() {
+		return this.waveDifficulty;
+	}
+
+	@Override
+	public int getWaveTimer() {
+		final long remainingTime = Math.multiplyFull(this.timeUntilNextWaveInSeconds, 1000) - (System.currentTimeMillis() - this.timeOfLastCheck);
+		if (remainingTime <= 0) return 0;
+		return Math.toIntExact(Math.ceilDiv(remainingTime, 1000));
+	}
+
+	@Override
+	public boolean addEntity(final IEntity entity) {
+		return this.enemyComposition.add(entity);
+	}
+
+	@Override
+	public boolean removeEntity(final IEntity entity) {
+		return this.enemyComposition.remove(entity);
+	}
+
+	@Override
+	public Collection<IEntity> getAllEntities() {
+		return Collections.unmodifiableCollection(this.enemyComposition);
+	}
+
+	@Override
+	public void nextWave() {
+		this.waveNumber++;
+		this.updateDifficulty();
+		this.enemyComposition = this.calculateWaveComposition();
+		this.assignSpawnLocation();
+		this.timeOfLastCheck = System.currentTimeMillis();
+	}
+
+	private void constructCatalog() {
+		ServiceLoader.load(EnemySPI.class).forEach(enemySPI -> {
+			EnemyRole currentRole = enemySPI.getRole();
+			if (!this.enemyCatalog.containsKey(currentRole)) {
+				this.enemyCatalog.put(currentRole, new ArrayList<>());
+			}
+			this.enemyCatalog.get(currentRole).add(enemySPI);
+		});
+	}
+
+	private void updateDifficulty() {
+		int baseDifficulty = 50;
+		double growthRate = 1.25;
```

Comment:
Might be relevant to store these as `private final` fields, and assign them in constructor. 
Then make overloaded constructor with default values for these properties. 
E.g.
```java
public WaveManager(final int startWaveNumber, final int waveCountdownInSeconds, final IGameSettings gameSettings) {
	this(startWaveNumber, waveCountdownInSeconds, gameSettings, 50, 1.25);
}

public WaveManager(final int startWaveNumber, final int waveCountdownInSeconds, final IGameSettings gameSettings, final int baseDifficulty, final double growthRate) {
	/* ... */
}

```

Reply:
can we take this up in the next meeting?